Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Definition and example of a coalgebra
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  Δ : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  ε : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ TensorProduct.map Δ .id ∘ₗ Δ =
+    TensorProduct.map .id Δ ∘ₗ Δ
+  /-- The counit satisfies the left counitality law -/
+  ε_id : TensorProduct.lid R A ∘ₗ TensorProduct.map ε .id ∘ₗ Δ = .id
+  /-- The counit satisfies the right counitality law -/
+  id_ε : TensorProduct.rid R A ∘ₗ TensorProduct.map .id ε ∘ₗ Δ = .id
+
+namespace Coalgebra
+
+section CommRingAddCommGroup
+
+variable {R : Type u} {A : Type v}
+variable [CommRing R] [AddCommGroup A] [Module R A] [Coalgebra R A]
+
+@[simp]
+theorem coassoc_apply (a : A) : TensorProduct.assoc R A A A (TensorProduct.map Δ .id (Δ a)) =
+    TensorProduct.map .id Δ (Δ a) := LinearMap.congr_fun coassoc a
+
+@[simp]
+theorem ε_id_apply (a : A) : TensorProduct.lid R A (TensorProduct.map ε .id (Δ a)) = a :=
+    LinearMap.congr_fun ε_id a
+
+@[simp]
+theorem id_ε_apply (a : A) : TensorProduct.rid R A (TensorProduct.map .id ε (Δ a)) = a :=
+    LinearMap.congr_fun id_ε a
+
+end CommRingAddCommGroup
+
+end Coalgebra
+
+open Coalgebra
+
+section CommRing
+
+variable (R : Type u) [CommRing R]
+
+namespace CommRing
+
+instance toCoalgebra : Coalgebra R R where
+  Δ := (TensorProduct.mk R R R) 1
+  ε := .id
+  coassoc := rfl
+  ε_id := by ext; simp
+  id_ε := by ext; simp
+
+@[simp]
+theorem Δ_apply (r : R) : Δ r = 1 ⊗ₜ[R] r := rfl
+
+@[simp]
+theorem ε_apply (r : R) : ε r = r := rfl
+
+end CommRing
+
+namespace Finsupp
+
+variable (S : Type v)
```

Comment:
A minor thing: normally we reserve `S` for rings or semirings. For arbitrary index types, we tend to prefer `ι` (and `i : ι` instead of `s : S` below):
```suggestion
variable (ι : Type v)
```